### ISSUE TITLE
Need of PyTorch Lightning based new Base Class 

### ISSUE THREAD

**Issue by codestryker** [deepchem/deepchem#2514]
## 🚀 Feature

I am proposing a design to integrate new base class that will ease the construction of PyTorch Lightning based models.
Here is the link of my design -> [[proposal](https://docs.google.com/document/d/1SDna7clPkTuWScANoufSXKfOlUN5tOKqigydlTaTlG0/edit?usp=sharing)]

## Motivation

PyTorch Lightning is more advantageous than PyTorch. It retains the flexibility of PyTorch by providing full control over key points in training. It abstracts out most of the PyTorch code. It allows to train our model on CPUs, GPUs, Multiple GPUs, or TPUs without changing a single line of  PyTorch code.

## Additional Context

Please give your views to improve the design.

**Comment by peastman** [comment]
Could you elaborate on what the goals of this proposal are, and what specific problems you're trying to solve?

We've had discussions about PyTorch Lightning in the past, and concluded that TorchModel and LightningModel are "equivalent" in the sense that they fill the same role.  Each of them wraps a torch.nn.Module and adds code for training, logging, etc.  They have different designs and features, so one might be preferable over the other in certain situations.  But we couldn't see any way it would make sense to combine them, because they largely provide alternate implementations of the same features.

Another important design consideration is that all DeepChem models need to provide a consistent API.  The public APIs for TorchModel and KerasModel are nearly identical.  To a user, it shouldn't matter whether the model they're using happens to be implemented with PyTorch or Keras.  Maintaining that consistency is important for any design.

**Comment by JainSamyak8840** [comment]
@peastman I am agreed with you that TorchModel and LightningModel are equivalent. As much as I understand that if we replaces TorchModel with LightningModel then it has to inherit both Model and pl.LightningModule class. Model class
has some abstract method like predict, evaluate etc so it need to be inherited. After that LightningModel must be directly
inherit in class like "LCNN" in lcnn module then be able to perform training on it. In lcnn module there is a class LCNNModel which is used to wrap up the model like instance of "LCNN" but the purpose of this class will end after the above approach.

According to my approach I was trying to utilize the Model class abstract method like fit, predict, evaluate etc that were
implemented in TorchModel and not providing by the lightning module and replacing the Torch Model's fit_generator functionality with LightningModel class. LightningModel will be inherit in class like "LCNN" then I wrap up it in LCNNModel and pass it to TorchModel class then I will perform training on the model by calling the pytorch lightning trainer.fit inside the fit function of TorchModel. Therefore I asked for new base class that is LightningModel.

I roughly implemented my approach and will only be able to decide whether code will be consistent or not after implementing it fully.

So, what should I do next?

**Comment by peastman** [comment]
Could you elaborate on the specific problems you're trying to solve?  For example, is there some specific feature you need that is provided by LightningModel and not by TorchModel?

**Comment by JainSamyak8840** [comment]
@peastman Sorry for slow response!. I am trying to add Multi GPU and TPU support as a specific feature which is not provided by TorchModel. TorchModel logs the loss after n training steps so I am doing the same for Multi GPU outputs in training_step_end function provided by PyTorch Lightning Module.
